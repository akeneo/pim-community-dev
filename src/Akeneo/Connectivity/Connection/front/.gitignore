# dependencies
/node_modules

# testing
/coverage
