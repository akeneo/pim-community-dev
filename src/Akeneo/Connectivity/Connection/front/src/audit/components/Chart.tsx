import React from 'react';
import {
    DomainTuple,
    VictoryAxis,
    VictoryChart,
    VictoryChartProps,
    VictoryLine,
    VictoryScatter,
    VictoryThemeDefinition,
} from 'victory';
import {grey, lightGreyStroke, purple} from '../event-chart-themes';

const yAxeTheme = {
    tickLabels: {
        fill: 'none',
    },
    axis: lightGreyStroke,
    grid: lightGreyStroke,
};
const daysAxeTheme = {
    tickLabels: {
        fontSize: 11,
        fontFamily: 'Lato',
        fontWeight: ({tickValue}: {tickValue: any}) => (7 === tickValue ? 'bold' : 'normal'),
        fill: ({tickValue}: {tickValue: any}) => (7 === tickValue ? purple : grey),
    },
    axis: {
        stroke: 'none',
    },
    grid: {
        stroke: 'none',
    },
};
const gridYAxesTheme = {
    tickLabels: {
        fill: 'none',
    },
    grid: {
        stroke: 'none',
    },
    axis: lightGreyStroke,
};

type ChartData = {x: number; y: number; xLabel: string; yLabel: string};
type Props = {
    data: ChartData[];
    theme: VictoryThemeDefinition;
    chartOptions?: VictoryChartProps;
};

export const Chart = ({data, theme, chartOptions}: Props) => {
    let yMax = data.reduce((maxY, {y}) => (y > maxY ? y : maxY), 0);
    let yMin = data.reduce((minY, {y}) => (y < minY ? y : minY), yMax);

    // If yMin & yMax are equals: arbitrary override yMax value to be superior than yMin.
    if (yMin === 0 && yMax === 0) {
        yMax = 1;
    }
    // yMax & yMin are differents from 0 (but equals): arbitrary center the value between yMin & yMax.
    if (yMax - yMin === 0) {
        yMax += 1;
        yMin -= 1;
    }

    // Define the chart visual domain, add a margin of 20% of the data values below yMin and above yMax.
    const yMaxDomain = yMax + (yMax - yMin) / 5;
    const yMinDomain = yMin - (yMax - yMin) / 5;

    // Define the absolute value used to draw each axe (total of 7).
    const step = (yMaxDomain - yMinDomain) / 7;
    const yGridAxes: number[] = [
        yMinDomain,
        yMinDomain + 1 * step,
        yMinDomain + 2 * step,
        yMinDomain + 3 * step,
        yMinDomain + 4 * step,
        yMinDomain + 5 * step,
        yMinDomain + 6 * step,
        yMaxDomain,
    ];

    // The first X value need to be hidden. It is used to see the hidden start of the chart.
    const xDaysValues = data.map(({xLabel}) => xLabel).slice(1);

    const yLabels = ({datum}: {datum: ChartData}) => datum.yLabel;

    const domain: {
        x: DomainTuple;
        y: DomainTuple;
    } = {
        x: [0.5, 7.5],
        y: [yMinDomain, yMaxDomain],
    };

    // The rendering order is based on the elements order. Axes must be first to be draw in background.
    return (
        <VictoryChart {...chartOptions} padding={0} domain={domain} theme={theme}>
            <VictoryAxis padding={0} offsetY={25} tickValues={xDaysValues} style={daysAxeTheme} />
            <VictoryAxis dependentAxis tickValues={yGridAxes} style={yAxeTheme} />
            {[1.5, 2.5, 3.5, 4.5, 5.5, 6.5, 7.5].map((value, index) => {
                return <VictoryAxis dependentAxis key={index} style={gridYAxesTheme} axisValue={value} />;
            })}
            <VictoryLine data={data} interpolation='monotoneX' />
            <VictoryScatter data={data} labels={yLabels} size={7} />
        </VictoryChart>
    );
};
