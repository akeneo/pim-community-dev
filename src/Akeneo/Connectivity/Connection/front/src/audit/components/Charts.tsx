import React, {useEffect} from 'react';
import {Connection} from '../../model/connection';
import {FlowType} from '../../model/flow-type.enum';
import {fetchResult} from '../../shared/fetch-result';
import {isOk} from '../../shared/fetch-result/result';
import {useRoute} from '../../shared/router';
import {sourceConnectionsFetched} from '../actions/dashboard-actions';
import {useDashboardDispatch, useDashboardState} from '../dashboard-context';
import {SourceConnection} from '../model/source-connection';
import {NoConnection} from './NoConnection';
import {UserSurvey} from './UserSurvey';
import {DataSourceCharts} from './DataSourceCharts';

export const Charts = () => {
    const dispatch = useDashboardDispatch();

    const route = useRoute('akeneo_connectivity_connection_rest_list');
    useEffect(() => {
        let cancelled = false;
        fetchResult<Connection[], never>(route).then(result => {
            if (isOk(result) && !cancelled) {
                const sourceConnections = result.value.filter(
                    (connection): connection is SourceConnection => FlowType.DATA_SOURCE === connection.flowType
                );

                dispatch(sourceConnectionsFetched(sourceConnections));
            }
        });
        return () => {
            cancelled = true;
        };
    }, [route, dispatch]);

    const state = useDashboardState();
    if (0 === Object.keys(state.sourceConnections).length) {
        return <NoConnection />;
    }

    return (
        <>
            <DataSourceCharts />
            <UserSurvey />
        </>
    );
};
