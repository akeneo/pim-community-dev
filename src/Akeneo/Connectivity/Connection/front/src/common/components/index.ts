import * as EmptyState from '../../common/components/EmptyState';
import * as Typography from '../../common/components/Typography';
import {ApplyButton} from './button/ApplyButton';
import {GreyButton} from './button/GreyButton';
import {GhostButton} from './button/GhostButton';
import {IconButton} from './button/IconButton';
import {ImportantButton} from './button/ImportantButton';
import {DropdownLink, SecondaryActionsDropdownButton} from './button/SecondaryActionsDropdownButton';
import {Checkbox} from './Checkbox';
import {Figure, FigureCaption, FigureImage} from './Figure';
import {Form} from './form/Form';
import {FormGroup} from './form/FormGroup';
import {FormInput} from './form/FormInput';
import {Helper, HelperLink, HelperTitle} from './Helper';
import {Modal} from './Modal';
import {PageContent} from './PageContent';
import {PageError} from './PageError';
import {PageHeader} from './PageHeader';
import {RuntimeError} from './RuntimeError';
import {Section} from './Section';
import {Select} from './select/Select';
import {Select2, Select2Configuration} from './Select2';
import {SmallHelper} from './SmallHelper';
import {Table, TableCell, TableHeaderCell, TableHeaderRow, TableRow} from './Table';
import {ToggleButton} from './ToggleButton';
import {Flag} from './Flag';
import {Loading} from './Loading';

export {
    ApplyButton,
    Checkbox,
    DropdownLink,
    Figure,
    FigureCaption,
    FigureImage,
    Form,
    FormGroup,
    FormInput,
    GreyButton,
    GhostButton,
    Helper,
    HelperLink,
    HelperTitle,
    IconButton,
    ImportantButton,
    Modal,
    PageContent,
    PageError,
    PageHeader,
    RuntimeError,
    SecondaryActionsDropdownButton,
    Section,
    Select,
    Select2,
    Select2Configuration,
    SmallHelper,
    Table,
    TableCell,
    TableHeaderCell,
    TableHeaderRow,
    TableRow,
    ToggleButton,
    EmptyState,
    Typography,
    Flag,
    Loading,
};
