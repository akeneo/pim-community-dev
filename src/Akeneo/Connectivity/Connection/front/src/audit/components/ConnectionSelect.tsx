import React from 'react';
import styled from 'styled-components';
import {Select} from '../../common';
import {PropsWithTheme} from '../../common/theme';
import {Connection} from '../../model/connection';
import {useMediaUrlGenerator} from '../../settings/use-media-url-generator';
import {Translate} from '../../shared/translate';

interface Props {
    connections: Connection[];
    onChange: (code?: string) => void;
}

export const ConnectionSelect = ({connections, onChange}: Props) => {
    const generate = useMediaUrlGenerator();

    const data = connections.reduce((data, connection) => {
        data[connection.code] = {
            label: connection.label,
            imageSrc: (connection.image && generate(connection.image, 'dropdown_select_picture')) || undefined,
        };
        return data;
    }, {} as {[code: string]: {label: string; imageSrc?: string}});

    return (
        <>
            <Label>
                <Translate id='akeneo_connectivity.connection.dashboard.connection_selector.title' />
            </Label>
            <Select
                data={data}
                onChange={onChange}
                dropdownTitle={<Translate id='akeneo_connectivity.connection.dashboard.connection_selector.title' />}
            />
        </>
    );
};

const Label = styled.span`
    color: ${({theme}: PropsWithTheme) => theme.color.grey140};
    font-size: ${({theme}: PropsWithTheme) => theme.fontSize.default};
    height: 44px;
    line-height: 44px;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
    ::after {
        content: ':';
        padding-right: 1ch;
    }
`;
