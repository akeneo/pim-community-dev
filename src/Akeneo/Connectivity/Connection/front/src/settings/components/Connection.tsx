import React from 'react';
import {useHistory} from 'react-router';
import {Figure, FigureCaption, FigureImage} from '../../common';
import imgUrl from '../../common/assets/illustrations/NewAPI.svg';
import {useMediaUrlGenerator} from '../use-media-url-generator';

interface Props {
    code: string;
    label: string;
    hasWrongCombination: boolean;
    image: string | null;
}

export const Connection = ({code, label, hasWrongCombination, image}: Props) => {
    const history = useHistory();
    const generateMediaUrl = useMediaUrlGenerator();

    return (
        <Figure onClick={() => history.push(`/connections/${code}/edit`)}>
            <FigureImage
                className='AknImage-display'
                src={null === image ? imgUrl : generateMediaUrl(image, 'thumbnail')}
                alt={label}
            />
            <FigureCaption title={label} warning={hasWrongCombination}>
                {label}
            </FigureCaption>
        </Figure>
    );
};
