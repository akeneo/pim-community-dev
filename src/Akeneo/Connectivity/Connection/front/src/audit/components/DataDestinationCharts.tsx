import React from 'react';
import {Loading} from '../../common';
import styled from '../../common/styled-with-theme';
import {FlowType} from '../../model/flow-type.enum';
import {Translate, useTranslate} from '../../shared/translate';
import {ConnectionSelect} from '../components/ConnectionSelect';
import {useDashboardState} from '../dashboard-context';
import useConnectionSelect from '../hooks/useConnectionSelect';
import {WeeklyAuditChart} from './Chart/WeeklyAuditChart';
import {NoConnection} from './NoConnection';
import {SectionTitle} from 'akeneo-design-system';

export const DataDestinationCharts = () => {
    const translate = useTranslate();

    const {connections, connectionCode, selectConnectionCode} = useConnectionSelect(FlowType.DATA_DESTINATION);
    const {events} = useDashboardState();

    if (0 === connections.filter(connection => connection.code !== '<all>').length) {
        return (
            <DataDestinationChartsContainer>
                <SectionTitle>
                    <SectionTitle.Title>
                        <Translate id='akeneo_connectivity.connection.dashboard.charts.outbound' />
                    </SectionTitle.Title>
                </SectionTitle>
                <NoConnectionContainer>
                    <NoConnection small flowType={FlowType.DATA_DESTINATION} />
                </NoConnectionContainer>
            </DataDestinationChartsContainer>
        );
    }

    return (
        <DataDestinationChartsContainer>
            <SectionTitle>
                <SectionTitle.Title>
                    <Translate id='akeneo_connectivity.connection.dashboard.charts.outbound' />
                </SectionTitle.Title>
                <SectionTitle.Spacer />
                <ConnectionSelect
                    connections={connections}
                    onChange={code => selectConnectionCode(code!)}
                    label={
                        <Translate id='akeneo_connectivity.connection.dashboard.connection_selector.title.destination' />
                    }
                />
            </SectionTitle>

            {events.product_read[connectionCode] ? (
                <WeeklyAuditChart
                    theme='green'
                    title={translate('akeneo_connectivity.connection.dashboard.charts.number_of_products_sent')}
                    weeklyAuditData={events.product_read[connectionCode]}
                />
            ) : (
                <Loading />
            )}
        </DataDestinationChartsContainer>
    );
};

const DataDestinationChartsContainer = styled.div`
    padding-bottom: 25px;
    display: block;
`;
const NoConnectionContainer = styled.div`
    border: 1px solid ${({theme}) => theme.color.grey60};
    padding-bottom: 20px;
    margin-top: 20px;
`;
