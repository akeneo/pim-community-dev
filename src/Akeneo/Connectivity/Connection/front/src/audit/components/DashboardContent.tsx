import React from 'react';
import {useDashboardState} from '../dashboard-context';
import {DataDestinationCharts} from './DataDestinationCharts';
import {DataSourceCharts} from './DataSourceCharts';
import {BusinessErrorCountWidget} from './ErrorManagement/BusinessErrorCountWidget';
import {DataSourceErrorChart} from './ErrorManagement/DataSourceErrorChart';
import {NoConnection} from './NoConnection';
import {UserSurvey} from './UserSurvey';

export const DashboardContent = () => {
    const state = useDashboardState();

    if (0 === Object.values(state.connections).length) {
        return <NoConnection />;
    }

    return (
        <>
            <DataSourceCharts />
            <DataDestinationCharts />
            <DataSourceErrorChart />
            <BusinessErrorCountWidget />
            <UserSurvey />
        </>
    );
};
