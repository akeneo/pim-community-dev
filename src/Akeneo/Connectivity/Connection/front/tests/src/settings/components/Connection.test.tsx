import {Connection} from '@src/settings/components/Connection';
import userEvent from '@testing-library/user-event';
import {createMemoryHistory} from 'history';
import * as React from 'react';
import {Router} from 'react-router';
import {act} from 'react-test-renderer';
import {renderWithProviders} from '../../../test-utils';

describe('Connection', () => {
    it('should redirect to the edit connection page when clicked', async () => {
        const history = createMemoryHistory();
        const {getByText} = renderWithProviders(
            <Router history={history}>
                <Connection
                    code={'google-shopping'}
                    label={'Google Shopping'}
                    image={'a/b/c/path.jpg'}
                    hasWrongCombination={false}
                />
            </Router>
        );

        await act(async () => {
            userEvent.click(getByText('Google Shopping'));

            return Promise.resolve();
        });

        expect(history.location.pathname).toBe('/connect/connection-settings/google-shopping/edit');
    });
});
