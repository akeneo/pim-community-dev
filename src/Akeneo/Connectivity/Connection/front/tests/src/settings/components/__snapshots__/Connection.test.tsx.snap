// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Connection should render 1`] = `
<div
  className="sc-htpNat cJouWi"
  onClick={[Function]}
>
  <div
    className="sc-EHOje ePjyrR"
  >
    <div
      className="sc-bxivhb cIYLov"
    >
      <img
        alt="Google Shopping"
        className="sc-ifAKCX cnjDHs AknImage-display"
        src="pim_enrich_media_show?filename=a%252Fb%252Fc%252Fpath.jpg&filter=thumbnail"
      />
    </div>
  </div>
  <div
    className="sc-bZQynM ckdTkM"
    title="Google Shopping"
  >
    Google Shopping
  </div>
</div>
`;
